### ISSUE TITLE
Connection, possible issues

### ISSUE THREAD

**Issue by anvilvapre** [OpenRA/OpenRA#20077]
Some may be serious. Other trivial. Some just comments. Let me know if you dislike them.

1.

https://github.com/OpenRA/OpenRA/blob/1969ae361c0f2e6633d516b837f82262cbc58870/OpenRA.Game/Network/Connection.cs#L120

NetworkConnection won't ever live shorter than it's connection threads are active (60+ seconds?)? Not any of the threads are cleaned up when the main thread terminates - i.e. due/after to some Exception in the main thread. In theory at some point a receiver thread could still be active on a past game.

1.b 

A thread is started - to directly after start multiple background threads to do the actual connect. This is likely to address the the servers locking up when clients drop. But this seems bad practice/patching to just start more threads.

1.c. 

In the past I believe I once noticed that Targets would return multilple entries for localhost. So 2 connection could get setup of which one then would get closed again. Since communication only starts after starting the receicing thread this can't do any harm - like the same player entering the lobby twice.

2.
 
https://github.com/OpenRA/OpenRA/blob/1969ae361c0f2e6633d516b837f82262cbc58870/OpenRA.Game/Network/Connection.cs#L155

It is not needed to copy endpoint into this thread? Target is readonly.  But  wouldn't endpoint have changed once the thread started or at least when the exception was thrown?

3.

Naming. Packet is officially a OSI layer 3/TCP thing. It's trivial on the one hand but also causes a lot of confusion about what a slow packet is and where delays come from and what sizes of a packet are. Perhaps use terminoly of HTTP i.e. Request and Response or Chunk. While this is a major update of the code perhaps worth while changing the naming.

4,

https://github.com/OpenRA/OpenRA/blob/1969ae361c0f2e6633d516b837f82262cbc58870/OpenRA.Game/Network/Connection.cs#L274

What happens if the the write operation is interrupted SocketError.Interrupted? I would assume a IOException with nested SocketException would occur Perhaps try a seperate try catch around all socket operations and optionally retry. But this code was also already there in the past.

5.

(Avoid LINQ.)
No validation of frame number?
Wouldn't i tbe just as simple to call ReceiveOrders multiple times with the same ack. Avoid the combining of orders.

https://github.com/OpenRA/OpenRA/blob/1969ae361c0f2e6633d516b837f82262cbc58870/OpenRA.Game/Network/Connection.cs#L334

5.b. 

Two if's both checking ackCount > 0.

https://github.com/OpenRA/OpenRA/blob/1969ae361c0f2e6633d516b837f82262cbc58870/OpenRA.Game/Network/Connection.cs#L325

6.

https://github.com/OpenRA/OpenRA/blob/1969ae361c0f2e6633d516b837f82262cbc58870/OpenRA.Game/Network/Connection.cs#L385

Only if it arrives. Clients often drop because of bad or lost connections. The tcp reset may be lost. I guess the Sync packets will make the server detect sooner of later that a client dropped.

7.

No try catch. Only cleanup upon success. Seems off.

https://github.com/OpenRA/OpenRA/blob/1969ae361c0f2e6633d516b837f82262cbc58870/OpenRA.Game/Network/Connection.cs#L191

7.

https://github.com/OpenRA/OpenRA/blob/1969ae361c0f2e6633d516b837f82262cbc58870/OpenRA.Game/Network/Connection.cs#L154

If connection setup to one target fails then `errorMessage` is set. It could be set at the same time by multiple threads. It is not (re)set when the connection setup is successfull.

Also the use of the class does not know if a connection setup is still in progress or if all connection attempts failed. Only by polling the errorMessage value it would know. The state is always NotConnected. 

Probably better for queue to contain a pair of (client, errorMessage) to also be also be able to return/indicate errors cleanly.

8.

Probably also better to call this right after TryTake. No wait to slowly start a receiver thread - new connections may be setup in the mean time.

https://github.com/OpenRA/OpenRA/blob/1969ae361c0f2e6633d516b837f82262cbc58870/OpenRA.Game/Network/Connection.cs#L188

9.
Should be a return? As reported in a comment earlier. 
https://github.com/OpenRA/OpenRA/blob/1969ae361c0f2e6633d516b837f82262cbc58870/OpenRA.Game/Network/Connection.cs#L82

10.

Never stopped. Assumes the server will one day close the connection? Will always then also log an exception and set error to connection failed? Shouldn't this be while (!disposing)?

https://github.com/OpenRA/OpenRA/blob/1969ae361c0f2e6633d516b837f82262cbc58870/OpenRA.Game/Network/Connection.cs#L206

**Comment by teinarss** [comment]
Yeah this code is a bit of a mess, One thing i can see is that we should add some cancellation tokens to abort the threads in a more controlled manner. Feel free to submit some PR around this.
